api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py

```diff
@@ -0,0 +1,65 @@
+from typing import Tuple, List
+
+import numpy as np
+import torch
+from app.pipelines import Pipeline
+from speechbrain.pretrained import SepformerSeparation, SpectralMaskEnhancement
+from huggingface_hub import HfApi
+
+
+class AudioToAudioPipeline(Pipeline):
+    def __init__(self, model_id: str):
+        info = HfApi().model_info(model_id)
+        tags = info.tags
+        if "audio-source-separation" in tags:
+            self.model = SepformerSeparation.from_hparams(source=model_id)
+            self.type = "audio-source-separation"
+        elif "speech-enhancement" in tags:
+            self.model = SpectralMaskEnhancement.from_hparams(source=model_id)
+            self.type = "speech-enhancement"
+        else:
+            raise ValueError(
+                "speechbrain needs to know if this model is `speech-enhancement` or `audio-source-separation`, please add one of this tag to enable inference"
+            )
+        self.sampling_rate = self.model.hparams.sample_rate
+
+    def __call__(self, inputs: np.array) -> Tuple[np.array, int, List[str]]:
+        """
+        Args:
+            inputs (:obj:`np.array`):
+                The raw waveform of audio received. By default sampled at `self.sampling_rate`.
+                The shape of this array is `T`, where `T` is the time axis
+        Return:
+            A :obj:`tuple` containing:
+              - :obj:`np.array`:
+                 The return shape of the array must be `C'`x`T'`
+              - a :obj:`int`: the sampling rate as an int in Hz.
+              - a :obj:`List[str]`: the annotation for each out channel.
+                    This can be the name of the instruments for audio source separation
+                    or some annotation for speech enhancement. The length must be `C'`.
+        """
+
+        if self.type == "speech-enhancement":
+            return self.enhance(inputs)
+        elif self.type == "audio-source-separation":
+            return self.separate(inputs)
+
+    def separate(self, inputs):
+        mix = torch.from_numpy(inputs)
+        est_sources = self.model.separate_batch(mix.unsqueeze(0))
+        est_sources = est_sources[0]
+
+        # C x T
+        est_sources = est_sources.transpose(1, 0)
+        # normalize for loudness
+        est_sources = est_sources / est_sources.abs().max(dim=1, keepdim=True).values
```

Comment:
Is there a reason to not do the normalization for Asteroid?

Reply:
Original code is by the authors of the library: https://github.com/huggingface/huggingface_hub/pull/76

I don't have any idea how desirable that is for all models in all scenarios. My very limited understanding of audio is that it's not really desirable to do that in general (any outlier value, which could very well be desired like a loud bang in music will make all the rest of the audio inaudible as far as I understand).

If it makes sense to it go generally maybe the wrapper `api-inference-community` could take care of it (or the widget).

Reply:
IMO the comment saying that we are "loudness" normalizing here is misleading. If I understand correctly the code normalizes each source by a constant factor (their maximum amplitude), which means that the "volume" of each source is changed (normalized so that the maximum is 1), but nothing about the actual contents are changed. This has nothing to do with those "loudness wars" etc. It just makes sure that you're less likely to have to change your computer volume when listening to each of the sources. (But it also means that the original "volume" information, if preserved by the neural network, is lost.)

Reply:
> Is there a reason to do the normalization only for Speechbrain and only for separation, not enhancement?

Probably because most separation models are trained with SI-SDR today, which can output enormous values, and this is less true for speech enhancement models. 

A good (better?) approach is to normalize the output to have the same max as the mixture, at least the scales are comparable if the model doesn't output garbage. 

> any outlier value, which could very well be desired like a loud bang in music will make all the rest of the audio inaudible

This is fully correct.

Reply:
If I may chime in here, I think it is okay to use the max from mixture. This being said, I haven't had any issues with the current normalization (I have tested on a sizeable number of examples.)

Reply:
I think I will keep your proposed normalization and we will update if needed.
Thanks.